Support for CLI output formatting arguments in 'get.py'

dmgav opened the pull request:
The proposed feature implements support for CLI arguments specifying output formatting for `caproto-get`:

- **floating point formats**: `-e -f -g -s -lx -lo -lb` 
- **integer numbers**: `-0x -0o -0b`

The arguments change output formatting whenever floating point or integer values are printed to `stdout`. If a format string is specified in the argument list (e.g. `--format="{pv_name} {response.data}"`), the changed formatting is applied to the field `{response.data}`. If the field specification differs (e.g. `{response.data[0]}`) then the default format or format supplied within `--format` string is used. For example, the argument sequence `-0b --format="{response.data[0]:X}"` will print integer as hex number, despite `-0b` argument that selects binary format. The formatting also works in `--wide` and `-terse` modes.   

EPICS caget accepts contradicting output formatting arguments and sets the format according to the last specified argument from the same group. Similar behavior is built into the proposed implementation. Parameters for float and int formats belong to different groups and are processed separately. For example, invocation of `caproto-get` (`simple.py` ioc):
```
caproto-get -f5 -0x -lx -0b simple:A simple:B
```
results in pv `simple:A` (int) to be printed in binary format and pv `simple:B` (float) to be rounded to the nearest int and printed in hex format.

**Custom separator**: `-F`. Sets custom separator for built-in or supplied format. The separator is inserted between fields and between array elements. Examples that set separator to `*` (consistent with the behavior of EPICS caget): `-F*`, `-F *`, `-F'*'`,`-F '*'`. EPICS caget accepts parameter -F without any value and sets separator to 's'. Such behavior is difficult to implement ('argparse' doesn't allow optional value), but it seems useless anyway. Also, EPICS caget accepts only single character separators, but the proposed implementation allows setting separator to a string of arbitrary length.
For example
``` 
caproto-get --a -F ' ** ' -e5 -0x simple:A simple:B simple:C
```
prints the pvs using standard `--wide` format string, setting separator between fields and elements of the array `simple:C` to `' ** '`. The floating point value of `simple:B` is printed in scientific format and int values of `simple:A` and `simple:B` are printed in hex format. 

The separator will also be inserted between fields if the format string is supplied with `--format` argument in case the format contains only spaces between the fields. For example if format string is specified as
```
-F ' ** ' --format="{pv_value}      {response.data[0]:X}" 
```
the separator will be inserted between two fields. If the definition is changed to
```
-F ' ** ' --format="{pv_value}      0x{response.data[0]:X}" 
```
by adding prefix `0x` before the second field, the separator will not be inserted. Note, that all spaces between `}` and `{` are replaced by the single separator: in the above example `}       {` is replaced by `} ** {`.

Added **support for standard EPICS type names** that start with `DBR_` (or `dbr_`). In EPICS `DBR_` prefix is optional and may be omitted. In the proposed implementation, if type name starts with `DBR_` (not case sensitive), the prefix is removed from the string before it is used for generating request. This feature improves compatibility with EPICS caget.

The **tests** were added for the proposed features. Unfortunately, the tests are basic and verify the exit codes returned by program in response to the sets of arguments, but not the formatting itself.

danielballan commented:
Wow, thanks for the contribution @dmgav! I look forward to reviewing this in detail in the next couple days, but in the meantime I want to express my gratitude for this useful, detail-oriented work.

klauer commented:
Seconding @danielballan's comment - thanks so much for the work here!

dmgav commented:
Thank you @danielballan and @klauer. I see that my submission does not pass AppVeyor test, but the reported errors are not related to the modules that I changed. Is the AppVeyor test up to date?

klauer commented:
AppVeyor can be safely ignored for now - it's not yet stable.

dmgav commented:
@danielballan, I believe that you did not have a chance to take a look at the code. I did a few more changes to the code and expanded support for the formatting arguments to `monitor.py`. I also moved the additional functions for processing of command arguments and output formatting to a separate module. As a result, the modified files `get.py` and `monitor.py` contain only minor changes compared with the original files. I also added a few tests for `monitor.py`.

I will upload the new version shortly, so that you could review all the changes at once when you have time.

danielballan commented on caproto/commandline/cli_print_formats.py:
```diff
@@ -0,0 +1,248 @@
+    return format_str
+
+
+def print_response_data(data=None, data_fmt=None):
```
Minor quibble: Reading the code where this is used, the name of this function momentarily misled me into thinking that it printed. Could this be named `format_response_data`?

dmgav replied:
You are right. I renamed the function  to `format_response_data`.

danielballan commented on caproto/commandline/get.py:
```diff
@@ -77,7 +82,56 @@ def main():
+        help=("Use <ofs> as an alternate output field separator (e.g. -F*, -F'*', -F '*', -F ' ** ')"))
+
     args = parser.parse_args()
+    # Remove contradicting format arguments. This function may be simply removed from code
```
Agreed that this is messy but worth keeping for the sake of compatibility with the reference implementation.

dmgav commented on caproto/commandline/cli_print_formats.py:
```diff
@@ -0,0 +1,248 @@
+
+    def is_set(self):
+        ''' Returns True if format string is not empty '''
+        return len(self.format) > 0
```
I removed `is_set()` from code. It can easily be reimplemented if needed.

dmgav commented:
Thank you for reviewing the code. I implemented the proposed minor changes and uploaded the corrected version.

danielballan commented:
Thanks for your patience with my slow review and for the fast turn-around. Great, well-scoped contribution!
